File: src/main/java/com/example/poc/RestClientService.java
package com.example.poc;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClient;

@Service
public class RestClientService {

    private final RestClient client;

    @Autowired
    public RestClientService(
            ContentInterceptor contentInterceptor,
            RestClient.Builder restClientBuilder,
            @Value("${my-client.base-url}")
            String baseUrl
    ) {

        this.client = restClientBuilder
                .baseUrl(baseUrl)
                .requestInterceptor(contentInterceptor)
                .build();
    }

    public String retrieveFoo() {
        return client
                .get()
                .uri("/foo")
                .retrieve()
                .body(String.class);
    }
}

File: src/main/kotlin/com/example/poc/RestClientService.kt
package com.example.poc

import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import org.springframework.web.client.RestClient

@Service
class RestClientService(
    contentInterceptor: ContentInterceptor,
    restClientBuilder: RestClient.Builder,
    @param:Value("\${my-client.base-url}") private val baseUrl: String,
) {
    private val client = restClientBuilder
        .baseUrl(baseUrl)
        // Register request interceptor consuming the response body twice
        .requestInterceptor(contentInterceptor)
        .build()

    fun retrieveFoo(): String? {
        return client
            .get()
            .uri("/foo")
            .retrieve()
            .body(String::class.java)
    }
}
